patches/server/1058-Properly-handle-experience-dropping-on-block-break.patch

```diff
@@ -0,0 +1,90 @@
++++ b/src/main/java/net/minecraft/world/level/block/Block.java
+@@ -333,23 +333,31 @@ public class Block extends BlockBehaviour implements ItemLike {
+             for (net.minecraft.world.item.ItemStack drop : net.minecraft.world.level.block.Block.getDrops(state, world.getMinecraftWorld(), pos, blockEntity)) {
+                 items.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(drop));
+             }
++            Block block = state.getBlock();
+             io.papermc.paper.event.block.BlockBreakBlockEvent event = new io.papermc.paper.event.block.BlockBreakBlockEvent(org.bukkit.craftbukkit.block.CraftBlock.at(world, pos), org.bukkit.craftbukkit.block.CraftBlock.at(world, source), items);
++            event.setExpToDrop(block.getExpDrop(state, (ServerLevel) world, pos, net.minecraft.world.item.ItemStack.EMPTY, true));
+             event.callEvent();
+             for (var drop : event.getDrops()) {
+                 popResource(world.getMinecraftWorld(), pos, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(drop));
+             }
+-            state.spawnAfterBreak(world.getMinecraftWorld(), pos, ItemStack.EMPTY, true);
++            state.spawnAfterBreak(world.getMinecraftWorld(), pos, ItemStack.EMPTY, false);
++            block.popExperience((ServerLevel) world, pos, event.getExpToDrop());
+         }
+         return true;
+     }
+     // Paper end
+ 
+     public static void dropResources(BlockState state, Level world, BlockPos pos, @Nullable BlockEntity blockEntity, @Nullable Entity entity, ItemStack tool) {
++    // Paper start
++        dropResources(state, world, pos, blockEntity, entity, tool, true);
++    }
++    public static void dropResources(BlockState state, Level world, BlockPos pos, @Nullable BlockEntity blockEntity, @Nullable Entity entity, ItemStack tool, boolean dropExperience) {
++    // Paper end
+         if (world instanceof ServerLevel) {
+             Block.getDrops(state, (ServerLevel) world, pos, blockEntity, entity, tool).forEach((itemstack1) -> {
+                 Block.popResource(world, pos, itemstack1);
+             });
+-            state.spawnAfterBreak((ServerLevel) world, pos, tool, true);
++            state.spawnAfterBreak((ServerLevel) world, pos, tool, dropExperience); // Paper
+         }
+ 
+     }
+@@ -429,11 +437,16 @@ public class Block extends BlockBehaviour implements ItemLike {
+         this.playerDestroy(world, player, pos, state, blockEntity, tool, true);
+     }
+     public void playerDestroy(Level world, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack tool, boolean includeDrops) {
++        // Paper end
++        // Paper start
++        this.playerDestroy(world, player, pos, state, blockEntity, tool, true, true);
++    }
++    public void playerDestroy(Level world, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack tool, boolean includeDrops, boolean dropExp) {
```

Comment:
This overload isn't used anywhere except right above it. And **if** we are adding another parameter to it... you need to change **all** of the overrides as well. If you are going to do that, just change the method we already add to have the extra parameter.